---
title: "Introduction to R"
subtitle: "Practice exercises - Answer key"
author: "Kim Dill-McFarland, <kadm@uw.edu>"
date: "version `r format(Sys.time(), '%B %d, %Y')`"
output:
  pdf_document:
    toc: yes
  html_document:
    toc: yes
    toc_float:
      collapsed: no
editor_options:
  chunk_output_type: console
---
```{r include=FALSE}
knitr::opts_chunk$set(fig.width = 4, fig.height = 3)
```

# Setup

Open the Intro R Rproject and start a new working script. Install any new packages, load packages, set a seed, and load data as we did in the workshop.

```{r}
#RNAseq expression and metadata
load("data/RSTR_data_clean_subset.RData")
```

# Exercises
## Day 1: Base R
### Project setup

1. What are the benefits of storing data in `RData` versus tables (`csv`, `tsv`, etc)?

* Several tables can be stored in a single object and can be loaded together in R with a single `load()` command
* Data are automatically compressed so they take up less hard-drive space
* Data formats are preserved in R such as factors, numbers you want to treat as characters, etc

2. Imagine a hypothetical project with the following data and results. How would you choose to setup your Rproject directory and sub-directories? This is something that may evolve over time, but it is helpful to start with a defined structure to make it easier for you and others to find things.
    - `.RData` file containing all cleaned data for the project
    - 2 `.csv` of raw RNAseq counts and sample metadata (what was cleaned to make the `.RData`)
    - 4 `.csv` with linear model results
    - 25 `.png` plots of gene expression, individual genes
    - 1 `.png` plot of gene expression, faceted with many genes
    - 2 `.R` scripts, 1 for linear modeling and 1 for making plots
    - 1 `.Rmd` report summarizing and interpreting the results
    
    
There are many options for this! I would do the following. Note that I personally like to use a lot of sub-directories.

```
project_name/
    data_clean/
        .RData
    data_raw/
        2 data .csv
    figs/
        genes/
            25 individual gene plot .png
        1 facetd gene plot .png
    results/
        4 linear model .csv
    scripts/
        2 .R scripts
    .Rmd
```

Another option.

```
project_name/
    data/
        .RData
        2 data .csv
    results/
        models/
            4 linear model .csv
        figs/
            25 individual gene plot .png
            1 facetd gene plot .png
    2 .R scripts
    .Rmd
```

### Data types

1. What is the difference between a `character` and `factor`?

* A character is any combination of alphanumeric (A-Z, 0-9) and other symbols (_ . - etc) that R treats like a single word. These are analogous to categorical variables in statistics. So for example, we have a variable with data on "MEDIA" vs "TB" in the workshop data set
* A factor is a character variable with some additional formatting. Factors have defined levels in a defined order. If you try to add data not of one of the defined levels, it is seen as an `NA` or missing. 

For example, we could format our MEDIA/TB variable to a factor

```{r}
factor(dat$targets$condition)
```

and force TB to be the first level even though it is the second alphabetically

```{r}
factor(dat$targets$condition, levels=c("TB","MEDIA"))
```

And if we only allow TB as a level, we can see the how R replaces everything else with `NA`

```{r}
factor(dat$targets$condition, levels=c("TB"))
```

2. What data type does R classify the date `2021.06`? What about `2021/06`? If it is not classified as a "date", how could this impact downstream analyses? Try to predict the outcomes before checking in R. 

* Both are treated as numeric. The first as a number with 2 decimal digits and the second as the result of 2021 divided by 6

```{r}
2021.06
class(2021.06)

2021/06
class(2021/06)
```

* This could dramatically impact results if these data were actually dates because 1) they are not being treated the same even though they are the same date, 2) scales will be wrong (as in 2021.06 is one month apart from 2021.07 but will be treated at 0.01 apart), 3) some functions that require a date won't run on a numeric, and other issues


2. Challenge: Checkout the package `lubridate` for functions to effectively work with dates in R.

* You can force date formatting like so and it is a lot more intuitive that base R's `as.Date()`. Here, you simply list which date components you have for year (y), month (m), and day (d) in the order they are in. That function does the rest!

```{r message=FALSE}
library(lubridate)

ym("2021.06")
ym("2021/06")

#And if we had a day
ymd("2021.06.2")
#A different order
mdy("06.2.2021")
```

* <https://lubridate.tidyverse.org/> has more on the `lubridate` package

3. You have an `S3` list object named `myData`and it contains 2 data frames named `A` and `B`. Within `B` there is a column named `variable1`. How do you access this variable?

```{r eval=FALSE}
myData$B$variable1
```

### Subsetting and filtering

Using `dat`:

1. What is the mean library size `lib.size`? 

```{r}
mean(dat$targets$lib.size)
```

2. Try running `summary(dat$targets)`. What kinds of data does it provide? Why are the results different for different variables?

```{r}
summary(dat$targets)
```

* You will get min, max, mean, and quartiles for numeric data
* You get the class and length of character vectors
* R automatically detects the data type and provides as much info as it can. Since character variables are simply words, this class has the least summary info.

3. How many libraries have a library size `lib.size` greater than 5 million and a normalization factor `norm.factors` less than 1?

```{r}
size.logical <- dat$targets$lib.size > 5E6
norm.logical <- dat$targets$norm.factors < 1

dat$targets[size.logical & norm.logical, ]

#Or combine it all together
dat$targets[dat$targets$lib.size > 5E6 & dat$targets$norm.factors < 1, ]

#And bonus, you can make R count the rows for you
nrow(dat$targets[size.logical & norm.logical, ])
```

4. Challenge: Using the function `grepl`, how many libraries are from a donor with an `RSID` that starts with "RS1025"?

```{r}
dat$targets[grepl("^RS1025", dat$targets$RSID), ]
```

* Note that `^` means the start of and `$` means the end of in a regular expression (regex) used in `grepl`

# R session

```{r}
sessionInfo()
```

***